python/src/iceberg/avro/reader.py

```diff
@@ -0,0 +1,318 @@
+
+
+class FloatReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> float:
+        return decoder.read_float()
+
+
+class DoubleReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> float:
+        return decoder.read_double()
+
+
+class DateReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> date:
+        return decoder.read_date_from_int()
+
+
+class TimeReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> time:
+        return decoder.read_time_micros_from_long()
+
+
+class TimestampReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> datetime:
+        return decoder.read_timestamp_micros_from_long()
+
+
+class TimestamptzReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> datetime:
+        return decoder.read_timestamp_micros_from_long()
+
+
+class StringReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> str:
+        return decoder.read_utf8()
+
+
+class UUIDReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> UUID:
+        return UUID(decoder.read_utf8())
+
+
+@dataclass(frozen=True)
+class FixedReader(Reader):
+    length: int = field()
+
+    def read(self, decoder: BinaryDecoder) -> bytes:
+        return decoder.read(self.length)
+
+
+class BinaryReader(Reader):
+    def read(self, decoder: BinaryDecoder) -> bytes:
+        return decoder.read_bytes()
+
+
+@dataclass(frozen=True)
+class DecimalReader(Reader):
+    precision: int = field()
+    scale: int = field()
+
+    def read(self, decoder: BinaryDecoder) -> Decimal:
+        return decoder.read_decimal_from_bytes(self.precision, self.scale)
+
+
+@dataclass(frozen=True)
+class OptionReader(Reader):
+    option: Reader = field()
+
+    def read(self, decoder: BinaryDecoder) -> Any | None:
+        # For the Iceberg spec it is required to set the default value to null
+        # From https://iceberg.apache.org/spec/#avro
+        # Optional fields must always set the Avro field default value to null.
+        #
+        # This means that null has to come first:
+        # https://avro.apache.org/docs/current/spec.html
+        # type of the default value must match the first element of the union.
+        # This is enforced in the schema conversion, which happens prior
+        # to building the reader tree
+        if decoder.read_int() > 0:
+            return self.option.read(decoder)
+        return None
+
+
+@dataclass(frozen=True)
+class StructReader(Reader):
+    fields: list[Reader] = field()
+
+    def read(self, decoder: BinaryDecoder) -> AvroStruct:
+        return AvroStruct([field.read(decoder) for field in self.fields])
+
+
+@dataclass(frozen=True)
+class ListReader(Reader):
+    element: Reader
+
+    def read(self, decoder: BinaryDecoder) -> list:
+        read_items = []
+        block_count = decoder.read_long()
+        while block_count != 0:
+            if block_count < 0:
+                block_count = -block_count
+                # We ignore the block size for now
```

Comment:
I think the intent is for `read` to always skip the block size. It is only used for data skipping.